File: internal/service/book_service.go
package service

import (
	"context"
	"regexp"
	"strings"
	"time"

	"bookmgr/internal/apperr"
	"bookmgr/internal/model"
	"bookmgr/internal/repository"
)

const (
	DefaultPageSize = 20
	MaxPageSize     = 100
)

var publishedDatePattern = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}$`)

// BookInput carries book fields as submitted by API/form clients, before
// validation and normalization into a model.Book.
type BookInput struct {
	Title         string
	Author        string
	Rating        *int
	Memo          *string
	ISBN          *string
	Publisher     *string
	PublishedDate *string
}

type Pagination struct {
	Page     int
	PageSize int
	Total    int
}

type BookService struct {
	repo repository.BookRepository
}

func NewBookService(repo repository.BookRepository) *BookService {
	return &BookService{repo: repo}
}

func (s *BookService) List(ctx context.Context, query string, page, pageSize int) ([]model.Book, Pagination, error) {
	if page < 1 {
		page = 1
	}
	if pageSize < 1 || pageSize > MaxPageSize {
		pageSize = DefaultPageSize
	}

	books, total, err := s.repo.List(ctx, query, page, pageSize)
	if err != nil {
		return nil, Pagination{}, err
	}
	return books, Pagination{Page: page, PageSize: pageSize, Total: total}, nil
}

func (s *BookService) Get(ctx context.Context, id int64) (*model.Book, error) {
	return s.repo.FindByID(ctx, id)
}

func (s *BookService) Create(ctx context.Context, input BookInput) (*model.Book, error) {
	book, err := validateInput(input)
	if err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, book); err != nil {
		return nil, err
	}
	return book, nil
}

func (s *BookService) Update(ctx context.Context, id int64, input BookInput) (*model.Book, error) {
	book, err := validateInput(input)
	if err != nil {
		return nil, err
	}
	book.ID = id
	if err := s.repo.Update(ctx, book); err != nil {
		return nil, err
	}
	return book, nil
}

func (s *BookService) Delete(ctx context.Context, id int64) error {
	return s.repo.Delete(ctx, id)
}

func validateInput(input BookInput) (*model.Book, error) {
	title := strings.TrimSpace(input.Title)
	if title == "" {
		return nil, apperr.NewValidationError("title is required")
	}
	if len([]rune(title)) > 255 {
		return nil, apperr.NewValidationError("title must be at most 255 characters")
	}

	author := strings.TrimSpace(input.Author)
	if author == "" {
		return nil, apperr.NewValidationError("author is required")
	}
	if len([]rune(author)) > 255 {
		return nil, apperr.NewValidationError("author must be at most 255 characters")
	}

	if input.Rating != nil && (*input.Rating < 1 || *input.Rating > 5) {
		return nil, apperr.NewValidationError("rating must be between 1 and 5")
	}

	memo := normalizeOptional(input.Memo)
	if memo != nil && len([]rune(*memo)) > 2000 {
		return nil, apperr.NewValidationError("memo must be at most 2000 characters")
	}

	isbn := normalizeOptional(input.ISBN)
	if isbn != nil {
		digits := strings.ReplaceAll(*isbn, "-", "")
		if !isAllDigits(digits) || (len(digits) != 10 && len(digits) != 13) {
			return nil, apperr.NewValidationError("isbn must be 10 or 13 digits (hyphens allowed)")
		}
	}

	publisher := normalizeOptional(input.Publisher)
	if publisher != nil && len([]rune(*publisher)) > 255 {
		return nil, apperr.NewValidationError("publisher must be at most 255 characters")
	}

	publishedDate := normalizeOptional(input.PublishedDate)
	if publishedDate != nil {
		if !publishedDatePattern.MatchString(*publishedDate) {
			return nil, apperr.NewValidationError("published_date must be in YYYY-MM-DD format")
		}
		if _, err := time.Parse("2006-01-02", *publishedDate); err != nil {
			return nil, apperr.NewValidationError("published_date must be a valid date")
		}
	}

	return &model.Book{
		Title:         title,
		Author:        author,
		Rating:        input.Rating,
		Memo:          memo,
		ISBN:          isbn,
		Publisher:     publisher,
		PublishedDate: publishedDate,
	}, nil
}

// normalizeOptional trims whitespace and converts an empty result to nil so
// optional fields are stored as SQL NULL rather than an empty string.
func normalizeOptional(s *string) *string {
	if s == nil {
		return nil
	}
	trimmed := strings.TrimSpace(*s)
	if trimmed == "" {
		return nil
	}
	return &trimmed
}

func isAllDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return true
}

File: server/app/services/book_service.py
import re
from dataclasses import dataclass
from datetime import date

from ..errors import ValidationError
from ..repositories.book_repository import BookRepository
from ..schemas import Book, BookInput

DEFAULT_PAGE_SIZE = 20
MAX_PAGE_SIZE = 100

_PUBLISHED_DATE_PATTERN = re.compile(r"^\d{4}-\d{2}-\d{2}$")


@dataclass
class Pagination:
    page: int
    page_size: int
    total: int


def _normalize_optional(value: str | None) -> str | None:
    if value is None:
        return None
    trimmed = value.strip()
    return trimmed or None


def _is_valid_isbn(digits: str) -> bool:
    if len(digits) == 13:
        return digits.isdigit()
    if len(digits) == 10:
        return digits[:9].isdigit() and (digits[9].isdigit() or digits[9] in "Xx")
    return False


def _validate_input(input: BookInput) -> Book:
    title = input.title.strip()
    if not title:
        raise ValidationError("title is required")
    if len(title) > 255:
        raise ValidationError("title must be at most 255 characters")

    author = input.author.strip()
    if not author:
        raise ValidationError("author is required")
    if len(author) > 255:
        raise ValidationError("author must be at most 255 characters")

    if input.rating is not None and not (1 <= input.rating <= 5):
        raise ValidationError("rating must be between 1 and 5")

    memo = _normalize_optional(input.memo)
    if memo is not None and len(memo) > 2000:
        raise ValidationError("memo must be at most 2000 characters")

    isbn = _normalize_optional(input.isbn)
    if isbn is not None and not _is_valid_isbn(isbn.replace("-", "")):
        raise ValidationError(
            "isbn must be 13 digits, or 10 digits with an optional trailing X "
            "check digit (hyphens allowed)"
        )

    publisher = _normalize_optional(input.publisher)
    if publisher is not None and len(publisher) > 255:
        raise ValidationError("publisher must be at most 255 characters")

    published_date = _normalize_optional(input.published_date)
    if published_date is not None:
        if not _PUBLISHED_DATE_PATTERN.match(published_date):
            raise ValidationError("published_date must be in YYYY-MM-DD format")
        try:
            date.fromisoformat(published_date)
        except ValueError as exc:
            raise ValidationError("published_date must be a valid date") from exc

    return Book(
        id=0,
        title=title,
        author=author,
        rating=input.rating,
        memo=memo,
        isbn=isbn,
        publisher=publisher,
        published_date=published_date,
        created_at="",
        updated_at="",
    )


class BookService:
    def __init__(self, repo: BookRepository):
        self._repo = repo

    def list(self, query: str, page: int, page_size: int) -> tuple[list[Book], Pagination]:
        if page < 1:
            page = 1
        if page_size < 1 or page_size > MAX_PAGE_SIZE:
            page_size = DEFAULT_PAGE_SIZE

        books, total = self._repo.list(query, page, page_size)
        return books, Pagination(page=page, page_size=page_size, total=total)

    def get(self, book_id: int) -> Book:
        return self._repo.find_by_id(book_id)

    def get_by_isbn(self, isbn: str) -> Book:
        isbn = isbn.strip()
        if not isbn:
            raise ValidationError("isbn is required")
        return self._repo.find_by_isbn(isbn)

    def create(self, input: BookInput) -> Book:
        book = _validate_input(input)
        return self._repo.create(book)

    def update(self, book_id: int, input: BookInput) -> Book:
        book = _validate_input(input)
        return self._repo.update(book_id, book)

    def delete(self, book_id: int) -> None:
        self._repo.delete(book_id)
